Private repos fail

I have a test repo that passes fine on public repos, but fails on private ones. Any help?

```
/usr/bin/git -c user.name=github-action-benchmark -c user.email=github@users.noreply.github.com -c http.https://github.com/.extraheader= fetch origin gh-pages:gh-pages
fatal: could not read Username for 'https://github.com': No such device or address
##[error]Command 'git' failed with args '-c user.name=github-action-benchmark -c user.email=github@users.noreply.github.com -c http.https://github.com/.extraheader= fetch origin gh-pages:gh-pages': fatal: could not read Username for 'https://github.com': No such device or address
: Error: The process '/usr/bin/git' failed with exit code 128
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+# Unreleased
+
+- **Fix:** Support private repositories (#105)
+
 <a name="v1.11.3"></a>
 # [v1.11.3](https://github.com/benchmark-action/github-action-benchmark/releases/tag/v1.11.3) - 31 Dec 2021
 

--- a/src/git.ts
+++ b/src/git.ts
@@ -80,3 +80,15 @@ export async function pull(token: string | undefined, branch: string, ...options
 
     return cmd(...args);
 }
+
+export async function fetch(token: string | undefined, branch: string, ...options: string[]): Promise<string> {
+    core.debug(`Executing 'git fetch' for branch '${branch}' with token and options '${options.join(' ')}'`);
+
+    const remote = token !== undefined ? getRemoteUrl(token) : 'origin';
+    let args = ['fetch', remote, `${branch}:${branch}`];
+    if (options.length > 0) {
+        args = args.concat(options);
+    }
+
+    return cmd(...args);
+}

--- a/src/write.ts
+++ b/src/write.ts
@@ -436,9 +436,9 @@ async function writeBenchmarkToGitHubPagesWithRetry(
 }
 
 async function writeBenchmarkToGitHubPages(bench: Benchmark, config: Config): Promise<Benchmark | null> {
-    const { ghPagesBranch, skipFetchGhPages } = config;
+    const { ghPagesBranch, skipFetchGhPages, githubToken } = config;
     if (!skipFetchGhPages) {
-        await git.cmd('fetch', 'origin', `${ghPagesBranch}:${ghPagesBranch}`);
+        await git.fetch(githubToken, ghPagesBranch);
     }
     await git.cmd('switch', ghPagesBranch);
     try {

--- a/test/git.spec.ts
+++ b/test/git.spec.ts
@@ -1,5 +1,5 @@
 import { deepStrictEqual as eq, notDeepStrictEqual as neq, strict as A } from 'assert';
-import { cmd, pull, push } from '../src/git';
+import { cmd, pull, push, fetch } from '../src/git';
 
 interface ExecOptions {
     listeners: {
@@ -173,6 +173,37 @@ describe('git', function () {
             ok(args);
             eq(args[0], 'git');
             eq(args[1], userArgs.concat(['pull', 'origin', 'my-branch', 'opt1', 'opt2']));
+        });
+    });
+
+    describe('fetch()', function() {
+        it('runs `git fetch` with given branch and options with token', async function() {
+            const stdout = await fetch('this-is-token', 'my-branch', 'opt1', 'opt2');
+            const args = fakedExec.lastArgs;
+
+            eq(stdout, 'this is test');
+            ok(args);
+            eq(args[0], 'git');
+            eq(
+                args[1],
+                userArgs.concat([
+                    'fetch',
+                    'https://x-access-token:this-is-token@github.com/user/repo.git',
+                    'my-branch:my-branch',
+                    'opt1',
+                    'opt2',
+                ]),
+            );
+        });
+
+        it('runs `git fetch` with given branch and options without token', async function() {
+            const stdout = await fetch(undefined, 'my-branch', 'opt1', 'opt2');
+            const args = fakedExec.lastArgs;
+
+            eq(stdout, 'this is test');
+            ok(args);
+            eq(args[0], 'git');
+            eq(args[1], userArgs.concat(['fetch', 'origin', 'my-branch:my-branch', 'opt1', 'opt2']));
         });
     });
 });

--- a/test/git.spec.ts
+++ b/test/git.spec.ts
@@ -176,8 +176,8 @@ describe('git', function () {
         });
     });
 
-    describe('fetch()', function() {
-        it('runs `git fetch` with given branch and options with token', async function() {
+    describe('fetch()', function () {
+        it('runs `git fetch` with given branch and options with token', async function () {
             const stdout = await fetch('this-is-token', 'my-branch', 'opt1', 'opt2');
             const args = fakedExec.lastArgs;
 
@@ -196,7 +196,7 @@ describe('git', function () {
             );
         });
 
-        it('runs `git fetch` with given branch and options without token', async function() {
+        it('runs `git fetch` with given branch and options without token', async function () {
             const stdout = await fetch(undefined, 'my-branch', 'opt1', 'opt2');
             const args = fakedExec.lastArgs;
 

--- a/test/write.spec.ts
+++ b/test/write.spec.ts
@@ -53,7 +53,7 @@ class FakedOctokit {
     }
 }
 
-type GitFunc = 'cmd' | 'push' | 'pull';
+type GitFunc = 'cmd' | 'push' | 'pull' | 'fetch';
 class GitSpy {
     history: [GitFunc, unknown[]][];
     pushFailure: null | string;
@@ -129,6 +129,10 @@ jest.mock('../src/git', () => ({
     },
     async pull(...args: unknown[]) {
         gitSpy.call('pull', args);
+        return '';
+    },
+    async fetch(...args: unknown[]) {
+        gitSpy.call('fetch', args);
         return '';
     },
 }));
@@ -904,7 +908,7 @@ describe('writeBenchmark()', function () {
             const autoPush = cfg.autoPush ?? true;
             const skipFetch = cfg.skipFetch ?? false;
             const hist: Array<[GitFunc, unknown[]] | undefined> = [
-                skipFetch ? undefined : ['cmd', ['fetch', 'origin', 'gh-pages:gh-pages']],
+                skipFetch ? undefined : ['fetch', [token, 'gh-pages']],
                 ['cmd', ['switch', 'gh-pages']],
                 fetch ? ['pull', [token, 'gh-pages']] : undefined,
                 ['cmd', ['add', path.join(dir, 'data.js')]],

